### ISSUE TITLE
IO::Handle setvbuf bug

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3523]
Migrated from [rt.perl.org#5971](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=5971) (status was 'open')

Searchable as RT5971$

**Comment by p5pRT** [comment]
### From me-01@ton.iguana.be



I Was doing some straces of perl's output&#8203;:

strace perl \-wle 'print STDERR "hoi"\, "wow"'

gives\(as expected\)&#8203;:

write\(2\, "hoi"\, 3\)                      = 3
write\(2\, "wow"\, 3\)                      = 3
write\(2\, "\\n"\, 1

Because I needed lines in my application\, I wanted line buffering\, so I
tried&#8203;:

strace perl \-wle 'use IO&#8203;::Handle '\_IOLBF'; STDERR\->setvbuf\($buffer\_var\, \_IOLBF\, 1024\); print STDERR "hoidyhu"\, "wowgrrr"'

which gives \(totally unexpected\)&#8203;:

write\(2\, "h"\, 1\)                        = 1
write\(2\, "oidyhu"\, 6\)                   = 6
write\(2\, "w"\, 1\)                        = 1
write\(2\, "owgrrr"\, 6\)                   = 6
write\(2\, "\\n"\, 1

However\, this&#8203;:

strace perl \-wle 'use IO&#8203;::Handle '\_IOLBF'; STDERR\->setvbuf\(my $buffer\_var\, \_IOLBF\, 1024\); print STDERR "hoidyhu"\, "wowgrrr"'

gives the expected&#8203;:

write\(2\, "hoidyhuwowgrrr\\n"\, 15\)        = 15

I think there's something very wrong if it matters if the buffer 
argument is a global or a lexical variable\. Things like that make 
me worry about what happens in memory\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium

Site configuration information for perl v5.6.0:

Configured by ton at Wed Aug  9 23:24:07 CEST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.0-test5, archname=i686-linux
    uname='linux quasar 2.4.0-test5 #2 sat jul 29 04:45:14 cest 2000 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=undef 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2 -fomit-frame-pointer', gccversion=egcs-2.91.66 19990314 (egcs-1.1.2 release), gccosandvers=
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccflags ='-fno-strict-aliasing -I/usr/local/include'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib -L/opt/gnu/lib'
    libpth=/usr/local/lib /opt/gnu/lib /lib /usr/lib
    libs=-lnsl -lndbm -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.1.so, so=so, useshrplib=true, libperl=libperl.so
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.6.0/i686-linux/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib -L/opt/gnu/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/lib/perl5/5.6.0/i686-linux
    /usr/lib/perl5/5.6.0
    /usr/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/lib/perl5/site_perl/5.6.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/ton
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ton/bin.Linux:/home/ton/bin:/home/ton/bin.SampleSetup:/usr/local/bin:/usr/local/sbin:/usr/local/jre/bin:/usr/local/jdk116_v5/bin:/home/oracle/product/8.1.5/bin:/usr/games/bin:/usr/X11R6/bin:/usr/share/bin:/usr/bin:/usr/sbin:/bin:/sbin:.
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @gannett-ggreer

This isn't "PerlIO" per se because the documentation notes&#8203;:

\- \- \- 8\< \- \- \- 8\< \- \- \-
WARNING&#8203;: The IO&#8203;::Handle&#8203;::setvbuf\(\) is not available by default on Perls
5\.8\.0 and later because setvbuf\(\) is rather specific to using the stdio
library\, while Perl prefers the new perlio subsystem instead\.
\- \- \- 8\< \- \- \- 8\< \- \- \-

Indeed on Linux I get&#8203;:
IO&#8203;::Handle&#8203;::setvbuf not implemented on this architecture at \-e line 1\.

\-\- 
George Greer

**Comment by p5pRT** [comment]
### From @Hugmeir

On Wed Jul 28 20&#8203;:27&#8203;:39 2010\, greerga wrote&#8203;:
> This isn't "PerlIO" per se because the documentation notes&#8203;:
> 
> \- \- \- 8\< \- \- \- 8\< \- \- \-
> WARNING&#8203;: The IO&#8203;::Handle&#8203;::setvbuf\(\) is not available by default on Perls
> 5\.8\.0 and later because setvbuf\(\) is rather specific to using the stdio
> library\, while Perl prefers the new perlio subsystem instead\.
> \- \- \- 8\< \- \- \- 8\< \- \- \-
> 
> Indeed on Linux I get&#8203;:
> IO&#8203;::Handle&#8203;::setvbuf not implemented on this architecture at \-e line 1\.

How would I need to compile Perl to test if this is still present?

**Comment by p5pRT** [comment]
### From @tonycoz

On Sat\, May 05\, 2012 at 04&#8203;:59&#8203;:02AM \-0700\, Brian Fraser via RT wrote&#8203;:
> On Wed Jul 28 20&#8203;:27&#8203;:39 2010\, greerga wrote&#8203;:
> > This isn't "PerlIO" per se because the documentation notes&#8203;:
> > 
> > \- \- \- 8\< \- \- \- 8\< \- \- \-
> > WARNING&#8203;: The IO&#8203;::Handle&#8203;::setvbuf\(\) is not available by default on Perls
> > 5\.8\.0 and later because setvbuf\(\) is rather specific to using the stdio
> > library\, while Perl prefers the new perlio subsystem instead\.
> > \- \- \- 8\< \- \- \- 8\< \- \- \-
> > 
> > Indeed on Linux I get&#8203;:
> > IO&#8203;::Handle&#8203;::setvbuf not implemented on this architecture at \-e line 1\.
> 
> How would I need to compile Perl to test if this is still present?

You'd have to build perl without perlio\, but Configure now always
enables perlio\.

Tony

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, May 08\, 2012 at 09&#8203;:54&#8203;:20AM \+1000\, Tony Cook wrote&#8203;:
> On Sat\, May 05\, 2012 at 04&#8203;:59&#8203;:02AM \-0700\, Brian Fraser via RT wrote&#8203;:
> > On Wed Jul 28 20&#8203;:27&#8203;:39 2010\, greerga wrote&#8203;:
> > > This isn't "PerlIO" per se because the documentation notes&#8203;:
> > > 
> > > \- \- \- 8\< \- \- \- 8\< \- \- \-
> > > WARNING&#8203;: The IO&#8203;::Handle&#8203;::setvbuf\(\) is not available by default on Perls
> > > 5\.8\.0 and later because setvbuf\(\) is rather specific to using the stdio
> > > library\, while Perl prefers the new perlio subsystem instead\.
> > > \- \- \- 8\< \- \- \- 8\< \- \- \-
> > > 
> > > Indeed on Linux I get&#8203;:
> > > IO&#8203;::Handle&#8203;::setvbuf not implemented on this architecture at \-e line 1\.
> > 
> > How would I need to compile Perl to test if this is still present?
> 
> You'd have to build perl without perlio\, but Configure now always
> enables perlio\.

"Now" as in post 5\.15\.0

If I build 5\.14\.0 like this&#8203;:

\./Configure \-de \-Uuseperlio && make \-j5 test\_prep

I see the same confusing difference&#8203;:

$ strace \-oout \-ewrite \./perl \-Ilib \-wle 'use IO&#8203;::Handle '\_IOLBF'; STDERR\->setvbuf\(my $buffer\_var\, \_IOLBF\, 1024\); print STDERR "hoidyhu"\, "wowgrrr"'
hoidyhuwowgrrr
$ cat out 
write\(2\, "hoidyhuwowgrrr\\n"\, 15\)        = 15

$ strace \-oout \-ewrite \./perl \-Ilib \-wle 'use IO&#8203;::Handle '\_IOLBF'; STDERR\->setvbuf\($buffer\_var\, \_IOLBF\, 1024\); print STDERR "hoidyhu"\, "wowgrrr"'
Name "main&#8203;::buffer\_var" used only once&#8203;: possible typo at \-e line 1\.
hoidyhuwowgrrr
$ cat out 
write\(2\, "Name \\"main&#8203;::buffer\_var\\" used onl"\.\.\.\, 68\) = 68
write\(2\, "h"\, 1\)                        = 1
write\(2\, "oidyhu"\, 6\)                   = 6
write\(2\, "w"\, 1\)                        = 1
write\(2\, "owgrrr"\, 6\)                   = 6
write\(2\, "\\n"\, 1\)                       = 1


Nicholas Clark

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, May 08\, 2012 at 02&#8203;:51&#8203;:51PM \+0100\, Nicholas Clark wrote&#8203;:
> On Tue\, May 08\, 2012 at 09&#8203;:54&#8203;:20AM \+1000\, Tony Cook wrote&#8203;:
> > On Sat\, May 05\, 2012 at 04&#8203;:59&#8203;:02AM \-0700\, Brian Fraser via RT wrote&#8203;:
> > > On Wed Jul 28 20&#8203;:27&#8203;:39 2010\, greerga wrote&#8203;:
> > > > This isn't "PerlIO" per se because the documentation notes&#8203;:
> > > > 
> > > > \- \- \- 8\< \- \- \- 8\< \- \- \-
> > > > WARNING&#8203;: The IO&#8203;::Handle&#8203;::setvbuf\(\) is not available by default on Perls
> > > > 5\.8\.0 and later because setvbuf\(\) is rather specific to using the stdio
> > > > library\, while Perl prefers the new perlio subsystem instead\.
> > > > \- \- \- 8\< \- \- \- 8\< \- \- \-
> > > > 
> > > > Indeed on Linux I get&#8203;:
> > > > IO&#8203;::Handle&#8203;::setvbuf not implemented on this architecture at \-e line 1\.
> > > 
> > > How would I need to compile Perl to test if this is still present?
> > 
> > You'd have to build perl without perlio\, but Configure now always
> > enables perlio\.
> 
> "Now" as in post 5\.15\.0
> 
> If I build 5\.14\.0 like this&#8203;:
> 
> \./Configure \-de \-Uuseperlio && make \-j5 test\_prep
> 
> I see the same confusing difference&#8203;:
> 
> $ strace \-oout \-ewrite \./perl \-Ilib \-wle 'use IO&#8203;::Handle '\_IOLBF'; STDERR\->setvbuf\(my $buffer\_var\, \_IOLBF\, 1024\); print STDERR "hoidyhu"\, "wowgrrr"'
> hoidyhuwowgrrr
> $ cat out 
> write\(2\, "hoidyhuwowgrrr\\n"\, 15\)        = 15
> 
> $ strace \-oout \-ewrite \./perl \-Ilib \-wle 'use IO&#8203;::Handle '\_IOLBF'; STDERR\->setvbuf\($buffer\_var\, \_IOLBF\, 1024\); print STDERR "hoidyhu"\, "wowgrrr"'
> Name "main&#8203;::buffer\_var" used only once&#8203;: possible typo at \-e line 1\.
> hoidyhuwowgrrr
> $ cat out 
> write\(2\, "Name \\"main&#8203;::buffer\_var\\" used onl"\.\.\.\, 68\) = 68
> write\(2\, "h"\, 1\)                        = 1
> write\(2\, "oidyhu"\, 6\)                   = 6
> write\(2\, "w"\, 1\)                        = 1
> write\(2\, "owgrrr"\, 6\)                   = 6
> write\(2\, "\\n"\, 1\)                       = 1

Ho ho ho\.

$ strace \-ofixed \-ewrite \./perl \-Ilib \-wle 'use IO&#8203;::Handle '\_IOLBF'; STDERR\->setvbuf\($buffer\_var\, \_IOLBF\, 1024\); print STDERR "hoidyhu"\, "wowgrrr"; 0 if $buffer\_var;'
hoidyhuwowgrrr
$ cat fixed 
write\(2\, "hoidyhuwowgrrr\\n"\, 15\)        = 15

From the C library man page&#8203;:

&nbsp;       Except  for unbuffered files\, the buf argument should point to a buffer
&nbsp;       at least size bytes long; this buffer will be used instead of the  cur\-
&nbsp;       rent buffer\.  If the argument buf is NULL\, only the mode is affected; a
&nbsp;       new buffer will be allocated on the next read or write operation\.   The
&nbsp;       setvbuf\(\)  function  may only be used after opening a stream and before
&nbsp;       any other operations have been performed on it\.


What's going wrong in the test case is that the used only once warning is
causing output to STDERR before setvbuf\(\) is called\, which violates the
calling constraints\.


IO&#8203;::Handle's documentation says\.

&nbsp;    You should only change the buffer before any I/O\, or immediately after
&nbsp;    calling flush\.


This actually seems to be erroneous\, as calling flush on the handle \*doesn't\*
make setvbuf "work"\.

$ strace \-obug \-ewrite \./perl \-Ilib \-wle 'use IO&#8203;::Handle '\_IOLBF'; STDERR\->flush or die; STDERR\->setvbuf\($buffer\_var\, \_IOLBF\, 1024\); print STDERR "hoidyhu"\, "wowgrrr";'
Name "main&#8203;::buffer\_var" used only once&#8203;: possible typo at \-e line 1\.
hoidyhuwowgrrr
$ cat bug
write\(2\, "Name \\"main&#8203;::buffer\_var\\" used onl"\.\.\.\, 68\) = 68
write\(2\, "h"\, 1\)                        = 1
write\(2\, "oidyhu"\, 6\)                   = 6
write\(2\, "w"\, 1\)                        = 1
write\(2\, "owgrrr"\, 6\)                   = 6
write\(2\, "\\n"\, 1\)                       = 1


Looks like a fix to IO&#8203;::Handle's documentation is needed\.

Also\, I guess that we should take this part out of its synopsis&#8203;:

&nbsp;    \# setvbuf is not available by default on Perls 5\.8\.0 and later\.
&nbsp;    use IO&#8203;::Handle '\_IOLBF';
&nbsp;    $io\->setvbuf\($buffer\_var\, \_IOLBF\, 1024\);

as it's really not useful any more\.

Nicholas Clark

**Comment by Leont** [comment]
I think we can close this as WONTFIX?

**Comment by tonycoz** [comment]
Yes, the setvbuf() being unavailable is documented so I think it's closable (and closed).
